### ISSUE TITLE
Implement Dirk Zimmer's Theta method

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8500]
The paper _Handling infinitely fast processes in continuous-time modelling_ by @dzimmer, presented at EOOLT '14 ([DOI:10.1145/2666202.2666206](https://doi.org/10.1145/2666202.2666206)), introduces a very nice and powerful concept to handle fast dynamic phenomena in equation-based, object-oriented models, so that the fast dynamics is exploited to ease the solution of the equations via a continuation method, but the obtained solution rigorously corresponds to that process being infinitely fast.

The particularly appealing property of that method is its user-friendliness: one just needs to use a special parameter Theta for the value of those process parameters that determine the speed of response of the very fast dynamics (i.e. small volumes, small capacitances or inductances, small inertias), without the need or rewriting or restructuring the code.

This method is particularly appealing to embed models of machines based on thermodynamic cycles (e.g. heat pumps, vapour compression machines, refrigeration cycles) into models with much slower dynamics, such as buildings, car interiors, aircraft cabins.

@AnHeuermann, this method could also be used to train ANNs that represent the infinitely fast behaviour in an even more efficient way, so I believe this is definitely relevant to your PhD research and the PhyMoS project.

This method should be implemented in OMC. It requires to introduce a special parameter. Rather than introducing a new reserved keyword in Modelica, which creates a number of issues, we could mark such parameter by an annotation, e.g. 

```Modelica
parameter Real Theta = 0 annotation(__OpenModelica_infinitelySmall);
```
which could later be standardized by the Modelica Association if the prototype implementation works well.

**Comment by AnHeuermann** [comment]
I think I got the idea of the paper, but not sure if I completely understood it in every detail.
Unfortunately Dirk Zimmer doesn't provide an example in his paper. I guess I need to check his other paper [Using Artificial States in Modeling Dynamic Systems: Turning Malpractice into Good Practice](https://ep.liu.se/ecp/084/009/ecp13084009.pdf) to find one.
But the idea to automatically handle idealized non-stiff systems instead of very stiff ones is interesting.

It seems to me that this is still an expert tool. One has to know what equations will result in such a hard to solve equation system and add `Theta` to all of the iteration variables. And of course the infinitely fast process has to be stable and have a steady-state solution.

The runtime implementation are probably straight forward. I would create a new non-linear solver that will first try to solve the steady-state problem with an implicit Euler and gives this solution as a start value to our usually used non-linear solver.
I'm not sure how many changes the index reduction needs to handle the `Theta` parameter at matching/index reduction and during tearing. @kabdelhak Any guess?

------------------

The connection to training ANNs I see is for training data generation. Making sure that the possibly hard to solve non-linear equation system is solvable by providing good initial guesses for an iterative (e.g. Newton) solver by first finding a steady-state solution for the "Theta" formulation of the system.

**Comment by casella** [comment]
> I guess I need to check his other paper [Using Artificial States in Modeling Dynamic Systems: Turning Malpractice into Good Practice](https://ep.liu.se/ecp/084/009/ecp13084009.pdf) to find one.

Yes. I can try to provide some simple examples myself. Maybe @dzimmer has some non-confidential example cases that he can provide.

> It seems to me that this is still an expert tool. One has to know what equations will result in such a hard to solve equation system

Definitely.

> and add `Theta` to all of the iteration variables.

No, that's the key point. You only have to use theta for those parameters that define the mass/momentum/energy/charge storage in that system. That's what makes it so powerful.

For example, an environmental control system such as the one described in the other paper you mention has fast dynamics coming from the inertia of the turbo-compressor and from the mass and energy storage in the three heat exchangers. You can make that dynamics infinitely fast by using theta for the shaft moment of inertia and for the volumes and masses of the heat exchangers. What is really nice is that you can take a physical model of the control system with existing dynamic models and turn that into a robust infinitely fast model just by setting a few parameters to Theta.

> And of course the infinitely fast process has to be stable and have a steady-state solution.

That goes without saying. But the whole point of this method is that the fast dynamics is actually real (or is an approximation of the real one), so it has a stable equilibrium by definition, otherwise the original system would blow up 😄 

> The connection to training ANNs I see is for training data generation. Making sure that the possibly hard to solve non-linear equation system is solvable by providing good initial guesses for an iterative (e.g. Newton) solver by first finding a steady-state solution for the "Theta" formulation of the system.

Yes. And eventually replace the whole thing with an explicit ANN model.

**Comment by dzimmer** [comment]
Hi
First of all, thanks for showing interest in the idea.

In the meantime, I've spent some time with these pseudo-dynamics and I found out that if you allow it in general, you can open up the gate to hell :-)

Anyway, let me show one way, so that you still have a useful solution but avoid hell. Allow Theta and perform index-reduction as usual. If after index-reduction, theta can be extracted as linear factor for a sub-set of derivatives, everything is fine and you can separate the time scales. All other cases represent an illegal use and must raise an error. 

You can then extract the code for the fast time-scale by the subset of affected state.

Then you have to go back to the original code and replace Theta by a symbolic zero and really make sure the zero does its job in multiplication (deleting entries in the structure incidence matrix). Perform index-reduction again. Then you have your slow dynamics. (and probably a difficult equation system where the fast system provides the solution)

If index-reduction fails, you have an error again. 

You may wonder where hell is: well it opens up as soon as you have as the equilibrium caused by Theta gives rise to further index-reduction. This might very well happen but then you have to rethink everything (which is actually quite interesting :-) )

This allows only a limited use of Theta but is better than nothing and can be implemented with reasonable effort.
There are multiple interesting ways forward but well a start is fine as well.

Use case: Well actually the ThermoFluid Stream Library contains a pseuodynamics for the inertia. Would be cool to idealize it away... Is maybe a complicated use case. But I can provide a very simplistic old version (without all the hustly of media-properties and the like)

**Comment by casella** [comment]
A first MWE for the method: [TestThetaMethod.mo.txt](https://github.com/OpenModelica/OpenModelica/files/10775352/TestThetaMethod.mo.txt)
